src/vm/methodtablebuilder.cpp

```diff
@@ -1517,12 +1518,20 @@ MethodTableBuilder::BuildMethodTableThrowing(
         if (hr == S_OK && (strcmp(nameSpace, "System.Runtime.Intrinsics.X86") == 0))
 #endif
         {
-            if (IsCompilationProcess())
+#if defined(CROSSGEN_COMPILE)
+#if defined(_TARGET_X86_) || defined(_TARGET_AMD64_)
+            if ((!IsNgenPDBCompilationProcess()
+                && GetAppDomain()->ToCompilationDomain()->GetTargetModule() != g_pObjectClass->GetModule())
+                || (strcmp(className, "Sse") != 0 && strcmp(className, "Sse2") != 0))
```

Comment:
We is this needed? Can we just not report them as unsupported in the compiler layer and have things work "naturally"?

Reply:
We need to avoid AOT compiling anything that touches the other intrinsics that could have different support level at runtime. It solves two problems:

1. The [reflection test](https://github.com/dotnet/coreclr/blob/b795318badbb2b7dd25715fff9f26153f261b4d4/tests/src/JIT/HardwareIntrinsics/X86/General/IsSupported.cs#L33-L52) that is trying to reflection-invoke e.g. `Avx.IsSupported` and finds out that the method body for `IsSupported` was crossgenned as returning `false`, but the reality is that Avx.IsSupported returns true. This could also be solved by not crossgenning any of the methods on the Avx type, but there's also problem number 2:
2. If there's an [intraprocedural dependency](https://github.com/dotnet/coreclr/blob/9773db1e7b1acb3ec75c9cc0e36bd62dcbacd6d5/src/System.Private.CoreLib/shared/System/Text/ASCIIUtility.cs#L83-L85), we'll hit bugs once tiering kicks in. Imagine that the linked method was calling `Avx.IsSupported` instead of `Sse2.IsSupported` - we crossgen it with `Avx.IsSupported == false`. At some point `GetIndexOfFirstNonAsciiByte` gets tiered up and `Avx.IsSupported == true`. At that point we start calling a method that doesn't have any guards for `Avx.IsSupported` and directly calls into intrinsics, but we pregenerated them as throwing `PlatformNotSupportedException`.